# U2A4 — Splits reproducibles de un dataset abierto

**Materia:** Ciberfísica e Inteligencia Artificial Aplicada  
**Nombre:** Esmeralda Gómez Huerta                                     
**Actividad:** U2A4 PRÁCTICA. Splits reproducibles de un dataset abierto           
**Dataset:** Wine Quality — vino tinto (UCI Machine Learning Repository)  
**Fuente:** https://archive.ics.uci.edu/dataset/186/wine+quality

---

## 1. ¿Qué son los splits de un dataset?

Un **split** (partición) consiste en dividir el dataset original en subconjuntos separados antes de entrenar un modelo de inteligencia artificial. Los dos más comunes son:

| Subconjunto | Proporción típica | Uso |
|---|---|---|
| **Train set** | 70% – 80% | El modelo aprende patrones con estos datos |
| **Test set** | 20% – 30% | Se evalúa el modelo con datos que nunca vio |

### ¿Por qué son importantes?

Si entrenamos y evaluamos con los **mismos datos**, el modelo puede simplemente memorizar las respuestas y obtener resultados artificialmente perfectos. Este problema se conoce como **sobreajuste (overfitting)**.

Al separar los datos:
- Medimos el desempeño **real** del modelo ante datos nuevos.
- Detectamos si el modelo generaliza bien o solo memoriza.
- Garantizamos una evaluación **objetiva e imparcial**.

### ¿Qué es la semilla fija (random_state)?

La división aleatoria puede producir resultados distintos cada vez que se ejecuta. Fijar una **semilla** (por ejemplo `random_state=42`) garantiza que siempre se genere exactamente la misma partición, haciendo el experimento **reproducible**: cualquier persona con el mismo código y dataset obtendrá los mismos archivos `train.csv` y `test.csv`.

### ¿Qué es stratify?

El parámetro `stratify=y` asegura que la proporción de cada clase (calidad del vino) se mantenga igual en train y test. Así evitamos que, por azar, una clase quede subrepresentada en algún subconjunto.

---

## 2. Importación de librerías

In [4]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import os

print('Librerías importadas correctamente ✓')
print(f'  pandas  : {pd.__version__}')
print(f'  numpy   : {np.__version__}')

Librerías importadas correctamente ✓
  pandas  : 3.0.2
  numpy   : 2.4.4


## 3. Carga del dataset

Cargamos el archivo **winequality-red.csv** desde la misma carpeta donde está este notebook.

In [5]:
# Cargar el dataset desde archivo local
df = pd.read_csv('winequality-red.csv', sep=';')

# Renombrar columnas para eliminar espacios
df.columns = [
    'fixed_acidity', 'volatile_acidity', 'citric_acid',
    'residual_sugar', 'chlorides', 'free_sulfur_dioxide',
    'total_sulfur_dioxide', 'density', 'pH',
    'sulphates', 'alcohol', 'quality'
]

print('Dataset cargado correctamente ✓')
print(f'Filas    : {df.shape[0]}')
print(f'Columnas : {df.shape[1]}')
df.head()

Dataset cargado correctamente ✓
Filas    : 1599
Columnas : 12


,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


## 4. Exploración básica del dataset

In [7]:
# Información general
print('=== Información del dataset ===')
df.info()
print()

# Verificar valores nulos
print('=== Valores nulos por columna ===')
print(df.isnull().sum())
print()

# Distribución de la variable objetivo
print('=== Distribución de calidad (quality) ===')
print(df['quality'].value_counts().sort_index())

=== Información del dataset ===
<class 'pandas.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed_acidity         1599 non-null   float64
 1   volatile_acidity      1599 non-null   float64
 2   citric_acid           1599 non-null   float64
 3   residual_sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free_sulfur_dioxide   1599 non-null   float64
 6   total_sulfur_dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB

=== Valores nulos por columna ===
fixed_acidity           0
volatile_acidity        0
citric_acid       

## 5. Partición train/test con semilla fija

In [8]:
# Separar características (X) y etiqueta (y)
X = df.drop('quality', axis=1)
y = df['quality']

# Parámetros de la partición
TEST_SIZE   = 0.2
RANDOM_SEED = 42

# Realizar la partición
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=TEST_SIZE,
    random_state=RANDOM_SEED,
    stratify=y
)

# Reconstruir DataFrames completos
train_df = X_train.copy()
train_df['quality'] = y_train.values

test_df = X_test.copy()
test_df['quality'] = y_test.values

# Resultados
print('=== Resultado de la partición ===')
print(f'Total original  : {len(df):>6} filas (100%)')
print(f'Train set       : {len(train_df):>6} filas ({len(train_df)/len(df)*100:.0f}%)')
print(f'Test set        : {len(test_df):>6} filas ({len(test_df)/len(df)*100:.0f}%)')
print()
print(f'Semilla utilizada (random_state): {RANDOM_SEED}')
print(f'Criterio de división: estratificado por calidad (stratify=y)')

=== Resultado de la partición ===
Total original  :   1599 filas (100%)
Train set       :   1279 filas (80%)
Test set        :    320 filas (20%)

Semilla utilizada (random_state): 42
Criterio de división: estratificado por calidad (stratify=y)


In [9]:
# Verificar que la distribución de clases se mantuvo proporcional
print('=== Proporción de calidad en cada split ===')
print()
print('Original:')
print((df['quality'].value_counts(normalize=True)*100).round(1).sort_index().to_string())
print()
print('Train:')
print((train_df['quality'].value_counts(normalize=True)*100).round(1).sort_index().to_string())
print()
print('Test:')
print((test_df['quality'].value_counts(normalize=True)*100).round(1).sort_index().to_string())

=== Proporción de calidad en cada split ===

Original:
quality
3     0.6
4     3.3
5    42.6
6    39.9
7    12.4
8     1.1

Train:
quality
3     0.6
4     3.3
5    42.6
6    39.9
7    12.4
8     1.2

Test:
quality
3     0.6
4     3.4
5    42.5
6    40.0
7    12.5
8     0.9


## 6. Guardar archivos en la carpeta /splits

In [10]:
# Crear carpeta splits/ si no existe
os.makedirs('splits', exist_ok=True)

# Guardar archivos CSV
train_df.to_csv('splits/train.csv', index=False)
test_df.to_csv('splits/test.csv',   index=False)

# Verificar que se guardaron correctamente
print('Archivos guardados en la carpeta splits/ ✓')
print()
for archivo in ['splits/train.csv', 'splits/test.csv']:
    verificacion = pd.read_csv(archivo)
    print(f'  {archivo}')
    print(f'    Filas    : {len(verificacion)}')
    print(f'    Columnas : {list(verificacion.columns)}')
    print()

Archivos guardados en la carpeta splits/ ✓

  splits/train.csv
    Filas    : 1279
    Columnas : ['fixed_acidity', 'volatile_acidity', 'citric_acid', 'residual_sugar', 'chlorides', 'free_sulfur_dioxide', 'total_sulfur_dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality']

  splits/test.csv
    Filas    : 320
    Columnas : ['fixed_acidity', 'volatile_acidity', 'citric_acid', 'residual_sugar', 'chlorides', 'free_sulfur_dioxide', 'total_sulfur_dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality']



## 7. Nota técnica

---

### ¿Qué son los splits y su importancia?

Los splits son particiones del dataset original en subconjuntos independientes. Su principal objetivo es separar los datos de **entrenamiento** (con los que el modelo aprende) de los datos de **prueba** (con los que se evalúa su desempeño real). Esta separación es fundamental para detectar problemas como el sobreajuste y para garantizar que la métrica de evaluación refleje la capacidad real del modelo de generalizar ante datos nuevos que nunca ha visto.

---

### Ficha técnica de la partición

| Campo | Detalle |
|---|---|
| **Nombre del dataset** | Wine Quality (vino tinto) |
| **Fuente original** | UCI Machine Learning Repository |
| **URL** | https://archive.ics.uci.edu/dataset/186/wine+quality |
| **Total de registros** | 1,599 filas |
| **Proporción utilizada** | 80% train / 20% test |
| **Semilla (random_state)** | 42 |
| **Criterio de división** | Estratificado por la variable `quality` (stratify=y) |
| **Archivos generados** | `splits/train.csv` (1,279 filas) / `splits/test.csv` (320 filas) |

---

### Créditos del dataset

> P. Cortez, A. Cerdeira, F. Almeida, T. Matos and J. Reis.  
> *Modeling wine preferences by data mining from physicochemical properties.*  
> In Decision Support Systems, Elsevier, 47(4):547-553, 2009.

Disponible en: UCI Machine Learning Repository — https://archive.ics.uci.edu/dataset/186/wine+quality